### PR TITLE
cluster: improve selfhosted cluster reliability

### PR DESCRIPTION
**Pull request by xiang90**
/cc @hongchaodeng

### CODE REVIEW

File: pkg/cluster/member.go [hunk 1]
```diff
@@ -30,6 +32,10 @@ func (c *Cluster) updateMembers(known etcdutil.MemberSet) error {
 	for _, m := range resp.Members {
 		var name string
 		if c.cluster.Spec.SelfHosted != nil {
+			if m.PeerURLs[0] == c.cluster.Spec.SelfHosted.BootMemberClientEndpoint {
+				return fmt.Errorf("skipping update members for self hosted cluster: waiting for the boot member to be removed...")
```
Comment (hongchaodeng) [inline pkg/cluster/member.go]: "boot member (%s)"

**Comment by hongchaodeng** [comment]
lgtm

File: pkg/cluster/self_hosted.go [hunk 2]
```diff
@@ -114,14 +114,16 @@ func (c *Cluster) migrateBootMember() error {
 		return err
 	}
 
-	delay := 20 * time.Second
-	c.logger.Infof("wait %v before removing the boot member", delay)
-	time.Sleep(delay)
+	go func() {
+		delay := 60 * time.Second
```
Comment (xiang90) [inline pkg/cluster/self_hosted.go]: @hongchaodeng I am not really sure if 60 is necessary. But let's keep it this way for safety? We can shorten this in the future if bootkube can work with shorter timeout?

File: pkg/cluster/member.go [hunk 3]
```diff
@@ -31,10 +33,16 @@ func (c *Cluster) updateMembers(known etcdutil.MemberSet) error {
 		var name string
 		if c.cluster.Spec.SelfHosted != nil {
 			name = m.Name
-			if len(name) == 0 {
+			if len(name) == 0 || len(m.ClientURLs) == 0 {
```
Comment (hongchaodeng) [inline pkg/cluster/member.go]: Why do we need to check ClientURL?

Comment (hongchaodeng) [inline pkg/cluster/member.go]: Oh I see. Boot member client endpoint.

Comment (xiang90) [inline pkg/cluster/member.go]: we are going to access m.ClientURL[0] later. we should check if the server finished publishing its client url. or we might panic.

File: pkg/cluster/self_hosted.go (hunk 2 above)
Comment (hongchaodeng) [inline pkg/cluster/self_hosted.go]: Add a TODO?

**Comment by hongchaodeng** [comment]
lgtm

**Pull request merged**